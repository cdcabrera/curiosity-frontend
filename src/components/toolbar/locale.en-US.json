{
  "curiosity-toolbar": {
    "button_displayName": "Search button for name",
    "clearFilters": "Reset filters",
    "label": "",
    "label_billing_provider": "",
    "label_billing_provider_none": "None",
    "label_billing_provider_aws": "Amazon Web Services",
    "label_billing_provider_azure": "Microsoft Azure",
    "label_billing_provider_gcp": "Google Cloud Platform",
    "label_billing_provider_oracle": "Oracle Cloud Infrastructure",
    "label_billing_provider_red hat": "Red Hat Marketplace",
    "label_category_cloud": "Public cloud",
    "label_category_hypervisor": "Hypervisor",
    "label_category_physical": "Physical",
    "label_category_virtual": "Virtual",
    "label_filter": "Filter by",
    "label_filter_billing_provider": "Purchased through",
    "label_filter_category": "Type",
    "label_filter_granularity": "Granularity",
    "label_filter_rangedMonthly": "Monthly",
    "label_filter_sla": "SLA",
    "label_filter_uom": "Unit of measure",
    "label_filter_usage": "Usage",
    "label_filter_variant": "Variant",
    "label_granularity": "",
    "label_granularity_Daily": "Past 30 days (daily)",
    "label_granularity_Weekly": "Past 12 weeks (weekly)",
    "label_granularity_Monthly": "Past year (monthly)",
    "label_granularity_Quarterly": "Past 3 years (quarterly)",
    "label_granularityRangedMonthly": "{{context}}",
    "label_granularityRangedMonthly_current": "This month",
    "label_groupVariant": "Variant:",
    "label_groupVariant_OpenShift Container Platform": "Container Platform (Annual)",
    "label_groupVariant_OpenShift-dedicated-metrics": "Dedicated (On-Demand)",
    "label_groupVariant_OpenShift-metrics": "Container Platform (On-Demand)",
    "label_groupVariant_RHEL for ARM": "RHEL ARM",
    "label_groupVariant_RHEL for IBM Power": "RHEL for IBM Power",
    "label_groupVariant_RHEL for IBM z": "RHEL for IBM z",
    "label_groupVariant_RHEL for x86": "RHEL for x86",
    "label_groupVariant_rhel-for-x86-eus": "RHEL for x86 EUS",
    "label_groupVariant_rhel-for-sap-x86": "RHEL for SAP x86",
    "label_groupVariant_rosa": "ROSA Hosted Control Planes",
    "label_groupVariant_Satellite Capsule": "Satellite Capsule",
    "label_groupVariant_Satellite Server": "Satellite Server",
    "label_sla": "",
    "label_sla_none": "No SLA",
    "label_sla_Premium": "Premium",
    "label_sla_Self-Support": "Self-Support",
    "label_sla_Standard": "Standard",
    "label_uom": "",
    "label_uom_none": "None",
    "label_uom_cores": "Cores",
    "label_uom_sockets": "Sockets",
    "label_usage": "",
    "label_usage_none": "Unspecified",
    "label_usage_Development/Test": "Development/Test",
    "label_usage_Disaster Recovery": "Disaster Recovery",
    "label_usage_Production": "Production",
    "placeholder": "Select",
    "placeholder_billing_provider": "Select purchased through",
    "placeholder_granularity": "Select date range",
    "placeholder_groupVariant": "Select a variant",
    "placeholder_rangedMonthly": "Select a month",
    "placeholder_sla": "Select SLA",
    "placeholder_uom": "Select unit",
    "placeholder_usage": "Select usage",
    "placeholder_filter": "Filter by",
    "placeholder_filter_billing_provider": "Filter by purchased through",
    "placeholder_filter_category": "Filter by type",
    "placeholder_filter_displayName": "Filter by name",
    "placeholder_filter_granularity": "Filter by date",
    "placeholder_filter_groupVariant": "Filter by variant",
    "placeholder_filter_rangedMonthly": "Filter by month",
    "placeholder_filter_sla": "Filter by SLA",
    "placeholder_filter_uom": "Filter by unit",
    "placeholder_filter_usage": "Filter by usage",
    "placeholder_filter_variant": "Filter by variant"
  }
}
