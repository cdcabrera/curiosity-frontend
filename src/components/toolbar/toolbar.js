import React from 'react';
import PropTypes from 'prop-types';
import {
  Toolbar as PfToolbar,
  ToolbarContent,
  ToolbarFilter,
  ToolbarGroup,
  ToolbarItem,
  ToolbarToggleGroup
} from '@patternfly/react-core';
import { FilterIcon } from '@patternfly/react-icons';
import _isEqual from 'lodash/isEqual';
import { connect, reduxTypes, store, useSelector } from '../../redux';
import { RHSM_API_QUERY_TYPES } from '../../types/rhsmApiTypes';
import { helpers } from '../../common';
import { translate } from '../i18n/i18n';
import { ToolbarFieldSla, toolbarFieldOptions as slaOptions } from './toolbarFieldSla';
import { ToolbarFieldUsage, toolbarFieldOptions as usageOptions } from './toolbarFieldUsage';
import { ToolbarFieldCategory, toolbarFieldOptions as categoryOptions } from './toolbarFieldCategory';

/**
 * Application filter toolbar.
 *
 * @augments React.Component
 * @fires onClearAllFilters
 * @fires onClearFilter
 */
/*
class Toolbar extends React.Component {
  componentDidUpdate(prevProps) {
    const { query } = this.props;
    console.log('onupdate >>>', _isEqual(query, prevProps.query));
    // if (!_isEqual(query, prevProps.query)) {
    // }
  }

  /**
   * Clear all filters' state.
   *
   * @event onClearAllFilters
   * /
  onClearAllFilters = () => {
    const { viewId } = this.props;

    store.dispatch({
      type: reduxTypes.query.SET_QUERY_CLEAR,
      viewId,
      clearFilters: {
        [RHSM_API_QUERY_TYPES.SLA]: null,
        [RHSM_API_QUERY_TYPES.USAGE]: null
      }
    });
  };

  /**
   * Clear individual filter state.
   *
   * @event onClearFilter
   * @param {string} categoryTitle
   * /
  onClearFilter = categoryTitle => {
    const { viewId } = this.props;
    const { value: categoryValue } = categoryOptions.find(({ title }) => title === categoryTitle) || {};

    if (!categoryValue) {
      return;
    }

    store.dispatch({
      type: reduxTypes.query.SET_QUERY_CLEAR,
      viewId,
      clearFilters: {
        [categoryValue]: null
      }
    });
  };

  /**
   * Return the current category filter, fallback to props selected.
   *
   * @returns {string|undefined}
   * /
  getCurrentCategoryFilter() {
    const { currentFilter, filterOptions } = this.props;

    return (
      currentFilter ||
      filterOptions.find(({ selected }) => selected === true)?.id ||
      (filterOptions.length === 1 && filterOptions[0]?.id)
    );
  }

  /**
   * Available, and selected select filter options for chip display.
   *
   * @param {string} field
   * @param {Array} options
   * @returns {Array}
   * /
  getSelectedFilterForChips(field, options) {
    const { query } = this.props;
    const filter = typeof query?.[field] === 'string' && options.find(({ value }) => value === query?.[field]);

    return (filter?.title && [filter.title]) || [];
  }

  /**
   * Render a filter node.
   *
   * @param {string} field
   * @returns {Node}
   * /
  renderFiltersWORKSISH(field) {
    const { t, viewId } = this.props;
    const currentCategoryFilter = this.getCurrentCategoryFilter();
    let selectedFilterForChips;
    let filter;

    switch (field) {
      case RHSM_API_QUERY_TYPES.USAGE:
        selectedFilterForChips = this.getSelectedFilterForChips(field, usageOptions);
        filter = <ToolbarFieldUsage viewId={viewId} />;
        break;
      case RHSM_API_QUERY_TYPES.SLA:
      default:
        selectedFilterForChips = this.getSelectedFilterForChips(field, slaOptions);
        filter = <ToolbarFieldSla viewId={viewId} />;
        break;
    }

    return (
      <ToolbarFilter
        key={field}
        chips={selectedFilterForChips}
        deleteChip={this.onClearFilter}
        categoryName={t('curiosity-toolbar.category', { context: field })}
        showToolbarItem={currentCategoryFilter === field}
      >
        {filter}
      </ToolbarFilter>
    );
  }

  renderFilters() {
    const { t, viewId } = this.props;
    const currentCategoryFilter = this.getCurrentCategoryFilter();

    return (
      <React.Fragment>
        <ToolbarFilter
          key={RHSM_API_QUERY_TYPES.SLA}
          chips={this.getSelectedFilterForChips(RHSM_API_QUERY_TYPES.SLA, slaOptions)}
          deleteChip={this.onClearFilter}
          categoryName={t('curiosity-toolbar.category', { context: RHSM_API_QUERY_TYPES.SLA })}
          // showToolbarItem={currentCategoryFilter === RHSM_API_QUERY_TYPES.SLA}
        >
          <ToolbarFieldSla viewId={viewId} />
        </ToolbarFilter>
        <ToolbarFilter
          key={RHSM_API_QUERY_TYPES.USAGE}
          chips={this.getSelectedFilterForChips(RHSM_API_QUERY_TYPES.USAGE, usageOptions)}
          deleteChip={this.onClearFilter}
          categoryName={t('curiosity-toolbar.category', { context: RHSM_API_QUERY_TYPES.USAGE })}
          // showToolbarItem={currentCategoryFilter === RHSM_API_QUERY_TYPES.USAGE}
        >
          <ToolbarFieldUsage viewId={viewId} />
        </ToolbarFilter>
      </React.Fragment>
    );
  }

  /**
   * Render toolbar category select.
   *
   * @returns {Node}
   * /
  renderCategories() {
    const { filterOptions, viewId } = this.props;
    const currentCategoryFilter = this.getCurrentCategoryFilter();
    const options = categoryOptions.filter(({ value }) => filterOptions.find(({ id }) => id === value));

    if (options.length > 1) {
      return (
        <ToolbarItem>
          <ToolbarFieldCategory
            viewId={viewId}
            options={options}
            value={currentCategoryFilter}
            toggleIcon={<FilterIcon />}
          />
        </ToolbarItem>
      );
    }

    return null;
  }

  /**
   * Render a filter toolbar.
   *
   * @returns {Node}
   * /
  render() {
    const { chipGroupSequence, filterOptions, isDisabled } = this.props;

    if (isDisabled) {
      return null;
    }

    // let updatedFilterOptions = filterOptions;

    /*
    if (chipGroupSequence.size) {
      updatedFilterOptions = [];
      Array.from(chipGroupSequence).forEach(field => {
        const filter = filterOptions.find(obj => obj.id === field);
        updatedFilterOptions.push(filter);
      });
    }
    * /

    return (
      <PfToolbar
        id="curiosity-toolbar"
        className="curiosity-toolbar pf-m-toggle-group-container ins-c-primary-toolbar"
        collapseListedFiltersBreakpoint="sm"
        clearAllFilters={this.onClearAllFilters}
      >
        <ToolbarContent>
          <ToolbarToggleGroup toggleIcon={<FilterIcon />} breakpoint="md">
            <ToolbarGroup variant="filter-group">
              {this.renderCategories()}
              {
                this.renderFilters()
                // updatedFilterOptions.map(({ id }) => this.renderFilters(id))
                // .sort(({ id: a }, { id: b }) => a && a.localeCompare(b))}
              }
            </ToolbarGroup>
          </ToolbarToggleGroup>
        </ToolbarContent>
      </PfToolbar>
    );
  }
}
*/

const Toolbar = ({ filterOptions, isDisabled, query, t, viewId }) => {
  const currentFilter = useSelector(({ toolbar }) => toolbar.filters?.[viewId]?.currentFilter, null);
  const updatedQuery = useSelector(({ view }) => view.query?.[viewId], query);

  if (isDisabled) {
    return null;
  }

  const updatedFilterOptions = categoryOptions.filter(({ value }) => filterOptions.find(({ id }) => id === value));
  const currentCategoryFilter =
    currentFilter ||
    filterOptions.find(({ selected }) => selected === true)?.id ||
    (filterOptions.length === 1 && filterOptions[0]?.id);

  const getSelectedFilterForChips = (field, options) => {
    const filter =
      typeof updatedQuery?.[field] === 'string' && options.find(({ value }) => value === updatedQuery?.[field]);
    return (filter?.title && [filter.title]) || [];
  };

  const onClearAllFilters = () =>
    store.dispatch({
      type: reduxTypes.query.SET_QUERY_CLEAR,
      viewId,
      clearFilters: {
        [RHSM_API_QUERY_TYPES.SLA]: null,
        [RHSM_API_QUERY_TYPES.USAGE]: null
      }
    });

  const onClearFilter = categoryTitle => {
    const { value: categoryValue } = categoryOptions.find(({ title }) => title === categoryTitle) || {};
    if (categoryValue) {
      store.dispatch({
        type: reduxTypes.query.SET_QUERY_CLEAR,
        viewId,
        clearFilters: {
          [categoryValue]: null
        }
      });
    }
  };

  return (
    <PfToolbar
      id="curiosity-toolbar"
      className="curiosity-toolbar pf-m-toggle-group-container ins-c-primary-toolbar"
      collapseListedFiltersBreakpoint="sm"
      clearAllFilters={onClearAllFilters}
    >
      <ToolbarContent>
        <ToolbarToggleGroup toggleIcon={<FilterIcon />} breakpoint="md">
          <ToolbarGroup variant="filter-group">
            {updatedFilterOptions.length > 1 && (
              <ToolbarItem>
                <ToolbarFieldCategory
                  viewId={viewId}
                  options={updatedFilterOptions}
                  value={currentCategoryFilter}
                  toggleIcon={<FilterIcon />}
                />
              </ToolbarItem>
            )}
            <ToolbarFilter
              key={RHSM_API_QUERY_TYPES.SLA}
              chips={getSelectedFilterForChips(RHSM_API_QUERY_TYPES.SLA, slaOptions)}
              deleteChip={onClearFilter}
              categoryName={t('curiosity-toolbar.category', { context: RHSM_API_QUERY_TYPES.SLA })}
              showToolbarItem={currentCategoryFilter === RHSM_API_QUERY_TYPES.SLA}
            >
              <ToolbarFieldSla viewId={viewId} />
            </ToolbarFilter>
            <ToolbarFilter
              key={RHSM_API_QUERY_TYPES.USAGE}
              chips={getSelectedFilterForChips(RHSM_API_QUERY_TYPES.USAGE, usageOptions)}
              deleteChip={onClearFilter}
              categoryName={t('curiosity-toolbar.category', { context: RHSM_API_QUERY_TYPES.USAGE })}
              showToolbarItem={currentCategoryFilter === RHSM_API_QUERY_TYPES.USAGE}
            >
              <ToolbarFieldUsage viewId={viewId} />
            </ToolbarFilter>
          </ToolbarGroup>
        </ToolbarToggleGroup>
      </ToolbarContent>
    </PfToolbar>
  );
};

/**
 * Prop types
 *
 * @type {{viewId: string, t: Function, activeFilters: Set, hardFilterReset: boolean, query: object,
 *     currentFilter: string, isDisabled: boolean, Array}}
 */
Toolbar.propTypes = {
  query: PropTypes.shape({
    [RHSM_API_QUERY_TYPES.SLA]: PropTypes.string,
    [RHSM_API_QUERY_TYPES.USAGE]: PropTypes.string
  }),
  // chipGroupSequence: PropTypes.instanceOf(Set),
  // activeFilters: PropTypes.instanceOf(Set),
  // currentFilter: PropTypes.oneOf([RHSM_API_QUERY_TYPES.SLA, RHSM_API_QUERY_TYPES.USAGE]),
  filterOptions: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.oneOf([RHSM_API_QUERY_TYPES.SLA, RHSM_API_QUERY_TYPES.USAGE]),
      selected: PropTypes.bool
    })
  ),
  // hardFilterReset: PropTypes.bool,
  isDisabled: PropTypes.bool,
  t: PropTypes.func,
  viewId: PropTypes.string
};

/**
 * Default props.
 *
 * @type {{viewId: string, t: translate, activeFilters: Set, hardFilterReset: boolean, query: object,
 *     currentFilter: string, isDisabled: boolean, filterOptions: Array}}
 */
Toolbar.defaultProps = {
  query: {},
  // activeFilters: new Set(),
  // chipGroupSequence: new Set(),
  // currentFilter: null,
  filterOptions: [
    {
      id: RHSM_API_QUERY_TYPES.SLA
    },
    {
      id: RHSM_API_QUERY_TYPES.USAGE,
      selected: true
    }
  ],
  // hardFilterReset: false,
  isDisabled: helpers.UI_DISABLED_TOOLBAR,
  t: translate,
  viewId: 'toolbar'
};

/**
 * Apply state to props.
 *
 * @param {object} state
 * @param {object} state.toolbar
 * @param {object} state.view
 * @param {object} props
 * @param {string} props.viewId
 * @returns {object}
 */
/*
const mapStateToProps = ({ toolbar, view }, { viewId }) => ({
  query: view?.query?.[viewId],
  ...toolbar.filters?.[viewId]
});

const ConnectedToolbar = connect(mapStateToProps)(Toolbar);

export { ConnectedToolbar as default, ConnectedToolbar, Toolbar };
 */
export { Toolbar as default, Toolbar };
