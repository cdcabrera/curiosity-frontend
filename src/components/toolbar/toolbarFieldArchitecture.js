import React from 'react';
import PropTypes from 'prop-types';
import { useHistory } from '../../hooks/useRouter';
import { useProduct } from '../productView/productViewContext';
import { reduxTypes, storeHooks } from '../../redux';
import { RHSM_API_QUERY_SET_TYPES } from '../../services/rhsm/rhsmConstants';
import { Select, SelectPosition } from '../form/select';
import { translate } from '../i18n/i18n';

/**
 * Generate select field options from config
 *
 * @param {object} options
 * @param {Function} options.useProduct
 * @returns {Function}
 */
const useToolbarFieldOptions = ({ useProduct: useAliasProduct = useProduct } = {}) => {
  const { productArchitectures } = useAliasProduct();
  const options = [];

  if (Array.isArray(productArchitectures)) {
    options.push(
      ...productArchitectures.map(type => ({
        title: translate('curiosity-toolbar.label', {
          context: ['architecture', (type === '' && 'unspecified') || type]
        }),
        value: type,
        selected: false
      }))
    );
  }

  return options;
};

/**
 * On select update.
 *
 * @param {object} options
 * @param {Function} options.useDispatch
 * @param {Function} options.useHistory
 * @param {Function} options.useProduct
 * @returns {Function}
 */
const useOnSelect = ({
  useDispatch: useAliasDispatch = storeHooks.reactRedux.useDispatch,
  useHistory: useAliasHistory = useHistory,
  useProduct: useAliasProduct = useProduct
} = {}) => {
  const { viewId } = useAliasProduct();
  const dispatch = useAliasDispatch();
  const history = useAliasHistory({ isSetAppNav: true });
  console.log('>>>', history);

  return ({ value = null } = {}) => {
    dispatch([
      {
        type: reduxTypes.query.SET_QUERY_RESET_INVENTORY_LIST,
        viewId
      },
      {
        type: reduxTypes.query.SET_QUERY,
        viewId,
        filter: RHSM_API_QUERY_SET_TYPES.ARCHITECTURE,
        value
      }
    ]);

    // history.push(value);
  };
};

/**
 * Display an architecture field with generated options.
 *
 * @fires onSelect
 * @param {object} props
 * @param {boolean} props.isFilter
 * @param {string} props.position
 * @param {Function} props.t
 * @param {Function} props.useOnSelect
 * @param {Function} props.useProduct
 * @param {Function} props.useToolbarFieldOptions
 * @returns {Node}
 */
const ToolbarFieldArchitecture = ({
  isFilter,
  position,
  t,
  useOnSelect: useAliasOnSelect,
  useProduct: useAliasProduct,
  useToolbarFieldOptions: useAliasToolbarFieldOptions
}) => {
  const { productId: updatedValue } = useAliasProduct();
  const onSelect = useAliasOnSelect();
  const options = useAliasToolbarFieldOptions();
  const updatedOptions = options.map(option => ({ ...option, selected: option.value === updatedValue }));

  return (
    <Select
      aria-label={t('curiosity-toolbar.placeholder', { context: [isFilter && 'filter', 'architecture'] })}
      onSelect={onSelect}
      options={updatedOptions}
      selectedOptions={updatedValue}
      placeholder={t('curiosity-toolbar.placeholder', { context: [isFilter && 'filter', 'architecture'] })}
      position={position}
      data-test="toolbarFieldArchitecture"
    />
  );
};

/**
 * Prop types.
 *
 * @type {{useOnSelect: Function, useProduct: Function, t: Function, isFilter: boolean, position: string,
 *     useToolbarFieldOptions: Function}}
 */
ToolbarFieldArchitecture.propTypes = {
  isFilter: PropTypes.bool,
  position: PropTypes.string,
  t: PropTypes.func,
  useOnSelect: PropTypes.func,
  useProduct: PropTypes.func,
  useToolbarFieldOptions: PropTypes.func
};

/**
 * Default props.
 *
 * @type {{useOnSelect: Function, useProduct: Function, t: translate, isFilter: boolean, position: string,
 *     useToolbarFieldOptions: Function}}
 */
ToolbarFieldArchitecture.defaultProps = {
  isFilter: false,
  position: SelectPosition.left,
  t: translate,
  useOnSelect,
  useProduct,
  useToolbarFieldOptions
};

export { ToolbarFieldArchitecture as default, ToolbarFieldArchitecture, useOnSelect, useToolbarFieldOptions };
