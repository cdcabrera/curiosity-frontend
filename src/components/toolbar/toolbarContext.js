import React from 'react';
import { ToolbarItem } from '@patternfly/react-core';
import { useProductQuery, useProductToolbarConfig } from '../productView/productViewContext';
import { RHSM_API_QUERY_SET_TYPES } from '../../services/rhsm/rhsmConstants';
import { useOnSelect as useSelectCategoryOnSelect, toolbarFieldOptions } from './toolbarFieldSelectCategory';
import { useOnSelect as useArchitectureOnSelect } from './toolbarFieldArchitecture';
import { useOnSelect as useBillingProviderOnSelect } from './toolbarFieldBillingProvider';
import { useOnSelect as useSlaOnSelect } from './toolbarFieldSla';
import { useOnSelect as useUsageOnSelect } from './toolbarFieldUsage';
import { SelectPosition } from '../form/select';
import { helpers } from '../../common/helpers';

/**
 * Clear a specific toolbar category using a select component's OnSelect hook.
 *
 * @param {object} options
 * @param {Function} options.useArchitectureOnSelect
 * @param {Function} options.useBillingProviderOnSelect
 * @param {Function} options.useSlaOnSelect
 * @param {Function} options.useUsageOnSelect
 * @returns {Function}
 */
const useToolbarFieldClear = ({
  useArchitectureOnSelect: useAliasArchitectureOnSelect = useArchitectureOnSelect,
  useBillingProviderOnSelect: useAliasBillingProviderOnSelect = useBillingProviderOnSelect,
  useSlaOnSelect: useAliasSlaOnSelect = useSlaOnSelect,
  useUsageOnSelect: useAliasUsageOnSelect = useUsageOnSelect
} = {}) => {
  const architectureOnSelect = useAliasArchitectureOnSelect();
  const billingOnSelect = useAliasBillingProviderOnSelect();
  const slaOnSelect = useAliasSlaOnSelect();
  const usageOnSelect = useAliasUsageOnSelect();

  return field => {
    switch (field) {
      case RHSM_API_QUERY_SET_TYPES.ARCHITECTURE:
        architectureOnSelect();
        break;
      case RHSM_API_QUERY_SET_TYPES.BILLING_PROVIDER:
        billingOnSelect();
        break;
      case RHSM_API_QUERY_SET_TYPES.SLA:
        slaOnSelect();
        break;
      case RHSM_API_QUERY_SET_TYPES.USAGE:
        usageOnSelect();
        break;
      default:
        break;
    }
  };
};

/**
 * Clear all available toolbar categories.
 *
 * @param {object} options
 * @param {Function} options.useProductQuery
 * @param {Function} options.useArchitectureOnSelect
 * @param {Function} options.useSelectCategoryOnSelect
 * @param {Function} options.useBillingProviderOnSelect
 * @param {Function} options.useSlaOnSelect
 * @param {Function} options.useUsageOnSelect
 * @returns {Function}
 */
const useToolbarFieldClearAll = ({
  useProductQuery: useAliasProductQuery = useProductQuery,
  useArchitectureOnSelect: useAliasArchitectureOnSelect = useArchitectureOnSelect,
  useSelectCategoryOnSelect: useAliasSelectCategoryOnSelect = useSelectCategoryOnSelect,
  useBillingProviderOnSelect: useAliasBillingProviderOnSelect = useBillingProviderOnSelect,
  useSlaOnSelect: useAliasSlaOnSelect = useSlaOnSelect,
  useUsageOnSelect: useAliasUsageOnSelect = useUsageOnSelect
} = {}) => {
  const {
    [RHSM_API_QUERY_SET_TYPES.ARCHITECTURE]: architecture,
    [RHSM_API_QUERY_SET_TYPES.BILLING_PROVIDER]: billingProvider,
    [RHSM_API_QUERY_SET_TYPES.SLA]: sla,
    [RHSM_API_QUERY_SET_TYPES.USAGE]: usage
  } = useAliasProductQuery();
  const architectureOnSelect = useAliasArchitectureOnSelect();
  const billingOnSelect = useAliasBillingProviderOnSelect();
  const slaOnSelect = useAliasSlaOnSelect();
  const usageOnSelect = useAliasUsageOnSelect();
  const selectCategoryOnSelect = useAliasSelectCategoryOnSelect();

  return hardFilterReset => {
    if (typeof architecture === 'string') {
      architectureOnSelect();
    }

    if (typeof billingProvider === 'string') {
      billingOnSelect();
    }

    if (typeof sla === 'string') {
      slaOnSelect();
    }

    if (typeof usage === 'string') {
      usageOnSelect();
    }

    if (hardFilterReset) {
      selectCategoryOnSelect();
    }
  };
};

/**
 * Return list of secondary toolbar fields for display.
 *
 * @param {object} options
 * @param {Array} options.categoryOptions
 * @param {Function} options.useProductToolbarConfig
 * @returns {Array}
 */
const useToolbarSecondaryFields = ({
  categoryOptions = toolbarFieldOptions,
  useProductToolbarConfig: useAliasProductToolbarConfig = useProductToolbarConfig
} = {}) => {
  const { secondaryFilters = [] } = useAliasProductToolbarConfig();

  return secondaryFilters.map(({ id, content }) => {
    const option = categoryOptions.find(({ value: categoryOptionValue }) => id === categoryOptionValue);
    const { component: OptionComponent } = option || {};

    return (
      (OptionComponent && (
        <ToolbarItem key={`option-${id}`}>
          <OptionComponent isFilter={false} position={SelectPosition.right} />
        </ToolbarItem>
      )) || <ToolbarItem key={helpers.generateId()}>{content}</ToolbarItem> ||
      null
    );
  });
};

const context = {
  useToolbarFieldClear,
  useToolbarFieldClearAll,
  useToolbarSecondaryFields
};

export { context as default, context, useToolbarFieldClear, useToolbarFieldClearAll, useToolbarSecondaryFields };
