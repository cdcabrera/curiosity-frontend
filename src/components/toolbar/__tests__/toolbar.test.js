import React from 'react';
import { mount, shallow } from 'enzyme';
import { Toolbar } from '../toolbar';
import { toolbarTypes } from '../toolbarTypes';
import { RHSM_API_QUERY_SLA, RHSM_API_QUERY_USAGE } from '../../../types/rhsmApiTypes';

describe('Toolbar Component', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should render a non-connected component', () => {
    const props = {};
    const component = shallow(<Toolbar {...props} />);

    expect(component).toMatchSnapshot('non-connected');
  });

  it('should handle updating state and dispatching a sla filter', () => {
    const props = {};

    const dispatchFilter = jest.spyOn(Toolbar.prototype, 'setDispatchFilter');
    const component = mount(<Toolbar {...props} />);
    component.setState({ filterCategory: RHSM_API_QUERY_SLA });

    const componentInstance = component.instance();

    const option = toolbarTypes.getOptions(RHSM_API_QUERY_SLA).options[0];
    componentInstance.onSlaSelect({ selected: { ...option }, value: option.value });
    expect(componentInstance.state).toMatchSnapshot(`${RHSM_API_QUERY_SLA} selected`);

    componentInstance.onClear();
    expect(componentInstance.state).toMatchSnapshot(`${RHSM_API_QUERY_SLA} clear filters`);

    expect(dispatchFilter).toHaveBeenCalledTimes(2);
  });

  it('should handle updating state and dispatching a usage filter', () => {
    const props = {};

    const dispatchFilter = jest.spyOn(Toolbar.prototype, 'setDispatchFilter');
    const component = mount(<Toolbar {...props} />);
    component.setState({ filterCategory: RHSM_API_QUERY_USAGE });

    const componentInstance = component.instance();

    const option = toolbarTypes.getOptions(RHSM_API_QUERY_USAGE).options[0];
    componentInstance.onUsageSelect({ selected: { ...option }, value: option.value });
    expect(componentInstance.state).toMatchSnapshot(`${RHSM_API_QUERY_USAGE} selected`);

    componentInstance.onClear();
    expect(componentInstance.state).toMatchSnapshot(`${RHSM_API_QUERY_USAGE} clear filters`);

    expect(dispatchFilter).toHaveBeenCalledTimes(2);
  });

  it('should return an empty render when disabled', () => {
    const props = {
      isDisabled: true
    };
    const component = shallow(<Toolbar {...props} />);

    expect(component).toMatchSnapshot('disabled component');
  });
});
