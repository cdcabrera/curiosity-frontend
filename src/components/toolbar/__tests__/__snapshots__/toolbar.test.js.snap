// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toolbar Component should handle updating state and dispatching a sla filter: sla clear filters 1`] = `
Object {
  "activeFilters": Set {},
  "filterCategory": null,
}
`;

exports[`Toolbar Component should handle updating state and dispatching a sla filter: sla selected 1`] = `
Object {
  "activeFilters": Set {
    "sla",
  },
  "filterCategory": "sla",
}
`;

exports[`Toolbar Component should handle updating state and dispatching a usage filter: usage clear filters 1`] = `
Object {
  "activeFilters": Set {},
  "filterCategory": null,
}
`;

exports[`Toolbar Component should handle updating state and dispatching a usage filter: usage selected 1`] = `
Object {
  "activeFilters": Set {
    "usage",
  },
  "filterCategory": "usage",
}
`;

exports[`Toolbar Component should render a non-connected component: non-connected 1`] = `
<Toolbar
  className="curiosity-toolbar pf-m-toggle-group-container ins-c-primary-toolbar"
  clearAllFilters={[Function]}
  collapseListedFiltersBreakpoint="sm"
  id="curiosity-toolbar"
>
  <ToolbarContent
    isExpanded={false}
    showClearFiltersButton={false}
  >
    <ToolbarToggleGroup
      breakpoint="md"
      toggleIcon={
        <FilterIcon
          color="currentColor"
          noVerticalAlign={false}
          size="sm"
        />
      }
    >
      <ForwardRef
        variant="filter-group"
      >
        <ToolbarItem>
          <Select
            aria-label="t(curiosity-toolbar.category)"
            ariaLabel="Select option"
            className=""
            id="generatedid-"
            isDisabled={false}
            isToggleText={true}
            name={null}
            onSelect={[Function]}
            options={
              Array [
                Object {
                  "title": "t(curiosity-toolbar.slaCategory)",
                  "value": "sla",
                },
                Object {
                  "title": "t(curiosity-toolbar.usageCategory)",
                  "value": "usage",
                },
              ]
            }
            placeholder="t(curiosity-toolbar.categoryPlaceholder)"
            selectedOptions={Array []}
            toggleIcon={
              <FilterIcon
                color="currentColor"
                noVerticalAlign={false}
                size="sm"
              />
            }
            variant="single"
          />
        </ToolbarItem>
        <ToolbarFilter
          categoryName="t(curiosity-toolbar.slaCategory)"
          chips={Array []}
          deleteChip={[Function]}
          showToolbarItem={false}
        >
          <Select
            aria-label="t(curiosity-toolbar.slaCategory)"
            ariaLabel="Select option"
            className=""
            id="generatedid-"
            isDisabled={false}
            isToggleText={true}
            name={null}
            onSelect={[Function]}
            options={
              Array [
                Object {
                  "title": "t(curiosity-toolbar.slaPremium)",
                  "value": "premium",
                },
                Object {
                  "title": "t(curiosity-toolbar.slaStandard)",
                  "value": "standard",
                },
                Object {
                  "title": "t(curiosity-toolbar.slaSelfSupport)",
                  "value": "self-support",
                },
                Object {
                  "title": "t(curiosity-toolbar.slaNone)",
                  "value": "",
                },
              ]
            }
            placeholder="t(curiosity-toolbar.slaPlaceholder)"
            selectedOptions={Array []}
            toggleIcon={null}
            variant="single"
          />
        </ToolbarFilter>
        <ToolbarFilter
          categoryName="t(curiosity-toolbar.usageCategory)"
          chips={Array []}
          deleteChip={[Function]}
          showToolbarItem={false}
        >
          <Select
            aria-label="t(curiosity-toolbar.usageCategory)"
            ariaLabel="Select option"
            className=""
            id="generatedid-"
            isDisabled={false}
            isToggleText={true}
            name={null}
            onSelect={[Function]}
            options={
              Array [
                Object {
                  "title": "t(curiosity-toolbar.usageDevelopment)",
                  "value": "Development/Test",
                },
                Object {
                  "title": "t(curiosity-toolbar.usageDisaster)",
                  "value": "Disaster Recovery",
                },
                Object {
                  "title": "t(curiosity-toolbar.usageProduction)",
                  "value": "Production",
                },
                Object {
                  "title": "t(curiosity-toolbar.usageUnspecified)",
                  "value": "",
                },
              ]
            }
            placeholder="t(curiosity-toolbar.usagePlaceholder)"
            selectedOptions={Array []}
            toggleIcon={null}
            variant="single"
          />
        </ToolbarFilter>
      </ForwardRef>
    </ToolbarToggleGroup>
  </ToolbarContent>
</Toolbar>
`;

exports[`Toolbar Component should return an empty render when disabled: disabled component 1`] = `""`;
